### ISSUE TITLE
`--allow-redefinition-new` does not mean "new" implementation but "new" variables

### ISSUE THREAD

**Issue by toslunar** [python/mypy#19918]
> https://github.com/python/mypy/blob/df05f05555ee62dbdb9960c64cad186172e92be1/docs/source/command_line.rst#L636-L639 
> https://github.com/python/mypy/blob/df05f05555ee62dbdb9960c64cad186172e92be1/docs/source/command_line.rst#L644-L646

The flag `--allow-redefinition-new` added by #18727 enhances inference of new variables, as the PR description says

> Infer union types for simple variables from multiple assignments, if the variable isn't annotated. 

This feature is rather orthogonal to the `--allow-redefinition` flag. In fact,

```
> mypy --version
mypy 1.18.2 (compiled: yes)

> mypy redef.py --allow-redefinition
Success: no issues found in 1 source file

> mypy redef.py --allow-redefinition-new --local-partial-types
redef.py:3: error: List comprehension has incompatible type List[list[str]]; expected List[str]  [misc]
Found 1 error in 1 file (checked 1 source file)
```

with the example in the doc of `--allow-redefinition`

```py
def process(items: list[str]) -> None:
    # 'items' has type list[str]
    items = [item.split() for item in items]
    # 'items' now has type list[list[str]]
```

**Comment by ilevkivskyi** [comment]
Hm @JukkaL should we imply `--allow-redefinition` semantics w.r.t. re-assigning a different type in `--allow-redefinition-new`?

**Comment by wyattscarpenter** [comment]
This documentation makes it very difficult to understand the present behavior of `--allow-redefinition-new` does, but based on the history of the feature (https://github.com/python/mypy/pull/18727, https://github.com/python/mypy/pull/19153, https://github.com/python/mypy/pull/19826) I think this is an feature bug not a docs issue, and the `--allow-redefinition-new` flag really does mean "new" implementation; and it really should implicitly enable `--allow-redefinition` semantics as well.

Thus this issue is the prophesied "create a GitHub issue if the new functionality doesn't support some important use cases."

**Comment by ilevkivskyi** [comment]
`new` here means new semantics. That said, the new semantics is supposed to be essentially a more lenient version of the old one (in particular allowing re-definition in multiple conditional branches). Taking into account we want to make this new semantics the default one for `--allow-redefinition` in v2.0 this is an important issue to fix soon.

**Comment by ilevkivskyi** [comment]
As was discussed before, there are various levels of "strength"  for `--allow-redefinition` w.r.t. explicit type annotations, see for example this comment https://github.com/python/mypy/issues/18516#issuecomment-2611001385. I am copying a modified version of that comment here:
```python
x = 1
x = [x]
reveal_type(x)
y: int = 1
y = [y]
reveal_type(y)
z: int = 1
z: list[int] = [z]
reveal_type(z)
t = 1
t: list[int] = [t]
reveal_type(t)
```
Note that for (current) `--allow-redefinition` to work one must _use_ a variable to allow redefinition. Current `--allow-redefinition` allows all four cases above, while `--allow-redefinition-new` allows only the first one.

If we want to completely replicate `--allow-redefinition` behavior with `--allow-redefinition-new` we would essentially allow all four without any restrictions on whether variable was used or not. Although some people may say it is more pythonic (myself included), I don't think we should do this. It is a big semantic change that some people may find controversial, and it likely will not help in many realistic use cases.

Now, if we want to simply enable use cases like the original example:
```python
def process(items: list[str]) -> None:
    items = [item.split() for item in items]
```
it would be sufficient to keep the current semantics of `--allow-redefinition-new` where we only allow redefinition for variables without explicit type annotation, but treat function arguments as inferred variables for the purpose of allowing redefinition.

We discussed with @JukkaL that this may need some care to avoid false positives, in particular, we will need to type-check r.h.s. twice, in known type context and in empty type context, and chose the attempt that doesn't cause errors. But it turns out this _already works_ in most cases, because we already do this (in some cases). For example this works:
```python
items = ["a",  "b", "c"]
reveal_type(items)  # list[str]
items = [item.split() for item in items]  # OK
reveal_type(items)  # list[list[str]]
```
as well as this one
```python
items = ["a",  "b", "c"]
reveal_type(items)  # list[str]
items = []  # OK
reveal_type(items) # list[str]
```
I tried some other examples, and the only problematic use case is when we combine both of the above, i.e. we do redefinition _twice_, where the second one is an empty list. FWIW this looks like something that should be easy to fix.
